File: couchbase-entity/src/main/java/com/kaufland/model/field/CblConstantHolder.java
package com.kaufland.model.field;

import com.kaufland.generation.TypeConversionMethodsGeneration;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.thoughtworks.qdox.model.JavaField;

import org.apache.commons.lang3.text.WordUtils;

import java.util.Arrays;
import java.util.List;

import javax.lang.model.element.Modifier;

import kaufland.com.coachbasebinderapi.Field;

/**
 * Created by sbra0902 on 21.06.17.
 */

public class CblConstantHolder extends CblBaseFieldHolder {

    private String mConstantValue;

    public CblConstantHolder(Field field) {
        super(field.name(), field);
        mConstantValue = field.defaultValue();
    }


    public String getConstantValue() {
        return mConstantValue;
    }

    @Override
    public MethodSpec getter(String dbName, boolean useMDocChanges) {
        TypeName returnType = TypeUtil.parseMetaType(getTypeMirror(), isIterable(), null);

        MethodSpec.Builder builder = MethodSpec.methodBuilder("get"  + accessorSuffix()).
                addModifiers(Modifier.PUBLIC).
                returns(returnType).
                addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDoc.get($N), $T.class)", getConstantName(), returnType);
        return builder.build();
    }

    @Override
    public MethodSpec setter(String dbName, TypeName entityTypeName, boolean useMDocChanges) {
        return null;
    }

    @Override
    public List<FieldSpec> createFieldConstant() {

        FieldSpec fieldAccessorConstant = FieldSpec.builder(String.class, getConstantName(), Modifier.FINAL, Modifier.PUBLIC, Modifier.STATIC).
                initializer("$S", getDbField()).
                build();

        return Arrays.asList(fieldAccessorConstant,
                FieldSpec.builder(String.class, "DOC_" + getConstantName(), Modifier.FINAL, Modifier.PUBLIC, Modifier.STATIC).
                        initializer("$S", getConstantValue()).
                        build());
    }
}

File: couchbase-entity/src/main/java/com/kaufland/model/field/CblConstantHolder.kt
package com.kaufland.model.field

import com.kaufland.generation.model.KDocGeneration
import com.kaufland.generation.model.TypeConversionMethodsGeneration
import com.kaufland.javaToKotlinType
import com.kaufland.model.deprecated.DeprecatedModel
import com.kaufland.util.ConversionUtil
import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.*

import java.util.Arrays

import kaufland.com.coachbasebinderapi.Field

/**
 * Created by sbra0902 on 21.06.17.
 */

class CblConstantHolder(field: Field) : CblBaseFieldHolder(field.name, field) {

    val constantValue: String = field.defaultValue

    val constantValueAccessorName = "DOC_$constantName"

    override val fieldType: TypeName = TypeUtil.parseMetaType(typeMirror, isIterable, null)

    override fun interfaceProperty(isOverride: Boolean): PropertySpec {
        val modifiers = listOfNotNull(KModifier.PUBLIC, KModifier.OVERRIDE.takeIf { isOverride })
        return PropertySpec.builder(accessorSuffix(), fieldType, modifiers).build()
    }

    override fun property(dbName: String?, possibleOverrides: Set<String>, useMDocChanges: Boolean, deprecated: DeprecatedModel?): PropertySpec {

        val builder = PropertySpec.builder(accessorSuffix(), fieldType, KModifier.PUBLIC, KModifier.OVERRIDE)
            .getter(FunSpec.getterBuilder().addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDoc.get(%N), %T::class)!!", constantName, fieldType).build())

        if (comment.isNotEmpty()) {
            builder.addKdoc(KDocGeneration.generate(comment))
        }

        return builder.build()
    }

    override fun createFieldConstant(): List<PropertySpec> {

        val fieldAccessorConstant = PropertySpec.builder(constantName, String::class, KModifier.FINAL, KModifier.PUBLIC).initializer("%S", dbField).addAnnotation(JvmField::class).build()

        return Arrays.asList(
            fieldAccessorConstant,
            PropertySpec.builder(constantValueAccessorName, typeMirror.asTypeName().javaToKotlinType(), KModifier.FINAL, KModifier.PUBLIC).initializer(ConversionUtil.convertStringToDesiredFormat(typeMirror, constantValue)).addAnnotation(JvmField::class).build()
        )
    }

    override fun builderSetter(dbName: String?, packageName: String, entitySimpleName: String, useMDocChanges: Boolean, deprecated: DeprecatedModel?): FunSpec? {
        return null
    }
}
